File: identity/src/test/java/com/android/identity/MobileSecurityObjectParserTest.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity;

import androidx.test.filters.SmallTest;

import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.security.PublicKey;
import java.security.cert.CertificateEncodingException;
import java.util.Map;
import java.util.Set;

import co.nstant.in.cbor.model.DataItem;

public class MobileSecurityObjectParserTest {

    @Test
    @SmallTest
    public void testMSOParserWithVectors() throws CertificateEncodingException {
        DataItem deviceResponse = Util.cborDecode(Util.fromHex(
                TestVectors.ISO_18013_5_ANNEX_D_DEVICE_RESPONSE));
        DataItem documentDataItem = Util.cborMapExtractArray(deviceResponse,
                "documents").get(0);

        DataItem issuerSigned = Util.cborMapExtractMap(documentDataItem, "issuerSigned");
        DataItem issuerAuthDataItem = Util.cborMapExtract(issuerSigned, "issuerAuth");

        DataItem mobileSecurityObjectBytes = Util.cborDecode(
                Util.coseSign1GetData(issuerAuthDataItem));
        DataItem mobileSecurityObject = Util.cborExtractTaggedAndEncodedCbor(
                mobileSecurityObjectBytes);
        byte[] encodedMobileSecurityObject = Util.cborEncode(mobileSecurityObject);

        // the response above and all the following constants are from ISO 18013-5 D.4.1.2 mdoc
        // response - the goal is to check that the parser returns the expected values
        MobileSecurityObjectParser.MobileSecurityObject mso = new MobileSecurityObjectParser()
                .setMobileSecurityObject(encodedMobileSecurityObject).parse();
        Assert.assertEquals("1.0", mso.getVersion());
        Assert.assertEquals("SHA-256", mso.getDigestAlgorithm());
        Assert.assertEquals("org.iso.18013.5.1.mDL", mso.getDocType());

        Assert.assertEquals(Set.of("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
                mso.getValueDigestNamespaces());
        Assert.assertNull(mso.getDigestIDs("abc"));
        Map<Long, byte[]> isoDigestIDs = mso.getDigestIDs("org.iso.18013.5.1");
        Assert.assertEquals(Set.of(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L, 11L, 12L),
                isoDigestIDs.keySet());
        Assert.assertEquals("75167333B47B6C2BFB86ECCC1F438CF57AF055371AC55E1E359E20F254ADCEBF"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(0L)));
        Assert.assertEquals("67E539D6139EBD131AEF441B445645DD831B2B375B390CA5EF6279B205ED4571"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(1L)));
        Assert.assertEquals("3394372DDB78053F36D5D869780E61EDA313D44A392092AD8E0527A2FBFE55AE"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(2L)));
        Assert.assertEquals("2E35AD3C4E514BB67B1A9DB51CE74E4CB9B7146E41AC52DAC9CE86B8613DB555"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(3L)));
        Assert.assertEquals("EA5C3304BB7C4A8DCB51C4C13B65264F845541341342093CCA786E058FAC2D59"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(4L)));
        Assert.assertEquals("FAE487F68B7A0E87A749774E56E9E1DC3A8EC7B77E490D21F0E1D3475661AA1D"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(5L)));
        Assert.assertEquals("7D83E507AE77DB815DE4D803B88555D0511D894C897439F5774056416A1C7533"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(6L)));
        Assert.assertEquals("F0549A145F1CF75CBEEFFA881D4857DD438D627CF32174B1731C4C38E12CA936"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(7L)));
        Assert.assertEquals("B68C8AFCB2AAF7C581411D2877DEF155BE2EB121A42BC9BA5B7312377E068F66"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(8L)));
        Assert.assertEquals("0B3587D1DD0C2A07A35BFB120D99A0ABFB5DF56865BB7FA15CC8B56A66DF6E0C"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(9L)));
        Assert.assertEquals("C98A170CF36E11ABB724E98A75A5343DFA2B6ED3DF2ECFBB8EF2EE55DD41C881"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(10L)));
        Assert.assertEquals("B57DD036782F7B14C6A30FAAAAE6CCD5054CE88BDFA51A016BA75EDA1EDEA948"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(11L)));
        Assert.assertEquals("651F8736B18480FE252A03224EA087B5D10CA5485146C67C74AC4EC3112D4C3A"
                .toLowerCase(), Util.toHex(isoDigestIDs.get(12L)));

        Map<Long, byte[]> isoUSDigestIDs = mso.getDigestIDs("org.iso.18013.5.1.US");
        Assert.assertEquals(Set.of(0L, 1L, 2L, 3L), isoUSDigestIDs.keySet());
        Assert.assertEquals("D80B83D25173C484C5640610FF1A31C949C1D934BF4CF7F18D5223B15DD4F21C"
                .toLowerCase(), Util.toHex(isoUSDigestIDs.get(0L)));
        Assert.assertEquals("4D80E1E2E4FB246D97895427CE7000BB59BB24C8CD003ECF94BF35BBD2917E34"
                .toLowerCase(), Util.toHex(isoUSDigestIDs.get(1L)));
        Assert.assertEquals("8B331F3B685BCA372E85351A25C9484AB7AFCDF0D2233105511F778D98C2F544"
                .toLowerCase(), Util.toHex(isoUSDigestIDs.get(2L)));
        Assert.assertEquals("C343AF1BD1690715439161ABA73702C474ABF992B20C9FB55C36A336EBE01A87"
                .toLowerCase(), Util.toHex(isoUSDigestIDs.get(3L)));

        PublicKey deviceKeyFromVector = Util.getPublicKeyFromIntegers(
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X, 16),
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y, 16));
        Assert.assertEquals(deviceKeyFromVector, mso.getDeviceKey());
        Assert.assertNull(mso.getDeviceKeyAuthorizedNameSpaces());
        Assert.assertNull(mso.getDeviceKeyAuthorizedDataElements());
        Assert.assertNull(mso.getDeviceKeyInfo());

        Assert.assertEquals(Timestamp.ofEpochMilli(1601559002000L), mso.getSigned());
        Assert.assertEquals(Timestamp.ofEpochMilli(1601559002000L), mso.getValidFrom());
        Assert.assertEquals(Timestamp.ofEpochMilli(1633095002000L), mso.getValidUntil());
        Assert.assertNull(mso.getExpectedUpdate());
    }
}

File: multipaz/src/commonTest/kotlin/org/multipaz/mdoc/mso/MobileSecurityObjectParserTest.kt
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.multipaz.mdoc.mso

import org.multipaz.cbor.Cbor
import org.multipaz.crypto.EcCurve
import org.multipaz.crypto.EcPublicKeyDoubleCoordinate
import org.multipaz.mdoc.TestVectors
import org.multipaz.util.fromHex
import org.multipaz.util.toHex
import kotlin.time.Instant
import org.multipaz.crypto.Algorithm
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class MobileSecurityObjectParserTest {
    @Test
    fun testMSOParserWithVectors() {
        val deviceResponse =
            Cbor.decode(TestVectors.ISO_18013_5_ANNEX_D_DEVICE_RESPONSE.fromHex())
        val documentDataItem = deviceResponse["documents"][0]
        val issuerSigned = documentDataItem["issuerSigned"]
        val issuerAuthDataItem = issuerSigned["issuerAuth"]
        val mobileSecurityObjectBytes = Cbor.decode(issuerAuthDataItem.asCoseSign1.payload!!)
        val mobileSecurityObject = mobileSecurityObjectBytes.asTaggedEncodedCbor
        val encodedMobileSecurityObject = Cbor.encode(mobileSecurityObject)

        // the response above and all the following constants are from ISO 18013-5 D.4.1.2 mdoc
        // response - the goal is to check that the parser returns the expected values
        val mso = MobileSecurityObjectParser(encodedMobileSecurityObject).parse()
        assertEquals("1.0", mso.version)
        assertEquals(Algorithm.SHA256, mso.digestAlgorithm)
        assertEquals("org.iso.18013.5.1.mDL", mso.docType)
        assertEquals(
            setOf("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
            mso.valueDigestNamespaces
        )
        assertNull(mso.getDigestIDs("abc"))
        val isoDigestIDs = mso.getDigestIDs("org.iso.18013.5.1")
        assertEquals(
            setOf(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L, 11L, 12L),
            isoDigestIDs!!.keys
        )
        assertEquals(
            "75167333b47b6c2bfb86eccc1f438cf57af055371ac55e1e359e20f254adcebf",
            isoDigestIDs[0L]!!.toHex()
        )
        assertEquals(
            "67e539d6139ebd131aef441b445645dd831b2b375b390ca5ef6279b205ed4571",
            isoDigestIDs[1L]!!.toHex()
        )
        assertEquals(
            "3394372ddb78053f36d5d869780e61eda313d44a392092ad8e0527a2fbfe55ae",
            isoDigestIDs[2L]!!.toHex()
        )
        assertEquals(
            "2e35ad3c4e514bb67b1a9db51ce74e4cb9b7146e41ac52dac9ce86b8613db555",
            isoDigestIDs[3L]!!.toHex()
        )
        assertEquals(
            "ea5c3304bb7c4a8dcb51c4c13b65264f845541341342093cca786e058fac2d59",
            isoDigestIDs[4L]!!.toHex()
        )
        assertEquals(
            "fae487f68b7a0e87a749774e56e9e1dc3a8ec7b77e490d21f0e1d3475661aa1d",
            isoDigestIDs[5L]!!.toHex()
        )
        assertEquals(
            "7d83e507ae77db815de4d803b88555d0511d894c897439f5774056416a1c7533",
            isoDigestIDs[6L]!!.toHex()
        )
        assertEquals(
            "f0549a145f1cf75cbeeffa881d4857dd438d627cf32174b1731c4c38e12ca936",
            isoDigestIDs[7L]!!.toHex()
        )
        assertEquals(
            "b68c8afcb2aaf7c581411d2877def155be2eb121a42bc9ba5b7312377e068f66",
            isoDigestIDs[8L]!!.toHex()
        )
        assertEquals(
            "0b3587d1dd0c2a07a35bfb120d99a0abfb5df56865bb7fa15cc8b56a66df6e0c",
            isoDigestIDs[9L]!!.toHex()
        )
        assertEquals(
            "c98a170cf36e11abb724e98a75a5343dfa2b6ed3df2ecfbb8ef2ee55dd41c881",
            isoDigestIDs[10L]!!.toHex()
        )
        assertEquals(
            "b57dd036782f7b14c6a30faaaae6ccd5054ce88bdfa51a016ba75eda1edea948",
            isoDigestIDs[11L]!!.toHex()
        )
        assertEquals(
            "651f8736b18480fe252a03224ea087b5d10ca5485146c67c74ac4ec3112d4c3a",
            isoDigestIDs[12L]!!.toHex()
        )
        val isoUSDigestIDs = mso.getDigestIDs("org.iso.18013.5.1.US")
        assertEquals(setOf(0L, 1L, 2L, 3L), isoUSDigestIDs!!.keys)
        assertEquals(
            "d80b83d25173c484c5640610ff1a31c949c1d934bf4cf7f18d5223b15dd4f21c",
            isoUSDigestIDs[0L]!!.toHex()
        )
        assertEquals(
            "4d80e1e2e4fb246d97895427ce7000bb59bb24c8cd003ecf94bf35bbd2917e34",
            isoUSDigestIDs[1L]!!.toHex()
        )
        assertEquals(
            "8b331f3b685bca372e85351a25c9484ab7afcdf0d2233105511f778d98c2f544",
            isoUSDigestIDs[2L]!!.toHex()
        )
        assertEquals(
            "c343af1bd1690715439161aba73702c474abf992b20c9fb55c36a336ebe01a87",
            isoUSDigestIDs[3L]!!.toHex()
        )
        val deviceKeyFromVector = EcPublicKeyDoubleCoordinate(
            EcCurve.P256,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X.fromHex(),
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y.fromHex()
        )
        assertEquals(deviceKeyFromVector, mso.deviceKey)
        assertNull(mso.deviceKeyAuthorizedNameSpaces)
        assertNull(mso.deviceKeyAuthorizedDataElements)
        assertNull(mso.deviceKeyInfo)
        assertEquals(Instant.fromEpochMilliseconds(1601559002000L), mso.signed)
        assertEquals(Instant.fromEpochMilliseconds(1601559002000L), mso.validFrom)
        assertEquals(Instant.fromEpochMilliseconds(1633095002000L), mso.validUntil)
        assertNull(mso.expectedUpdate)
    }
}
